server/master/src/main/java/org/apache/accumulo/master/metrics/fate/FateMetrics.java

```diff
@@ -16,61 +16,234 @@
+
   private final AtomicReference<FateMetricValues> metricValues;
 
   private volatile long lastUpdate = 0;
 
+  private final IZooReaderWriter zooReaderWriter;
+  private final ReadOnlyTStore<FateMetrics> zooStore;
+  private final String fateRootPath;
+
   public FateMetrics(final ServerContext context, final long minimumRefreshDelay) {
     super("Fate", "Fate Metrics", "fate");
 
-    this.context = context;
+    zooReaderWriter = context.getZooReaderWriter();
+    fateRootPath = context.getZooKeeperRoot() + Constants.ZFATE;
+
+    try {
+
+      zooStore = new ZooStore<>(fateRootPath, zooReaderWriter);
+
+    } catch (KeeperException ex) {
+      throw new IllegalStateException(
+          "FATE Metrics - Failed to create zoo store - metrics unavailable", ex);
+    } catch (InterruptedException ex) {
+      Thread.currentThread().interrupt();
+      throw new IllegalStateException(
+          "FATE Metrics - Interrupt received while initializing zoo store");
+    }
 
     this.minimumRefreshDelay = Math.max(DEFAULT_MIN_REFRESH_DELAY, minimumRefreshDelay);
 
-    metricValues = new AtomicReference<>(FateMetricValues.updateFromZookeeper(context, null));
+    metricValues = new AtomicReference<>(updateFromZookeeper());
 
     MetricsRegistry registry = super.getRegistry();
+
     currentFateOps = registry.newGauge("currentFateOps", "Current number of FATE Ops", 0L);
     zkChildFateOpsTotal = registry.newGauge("totalFateOps", "Total FATE Ops", 0L);
     zkConnectionErrorsTotal =
         registry.newGauge("totalZkConnErrors", "Total ZK Connection Errors", 0L);
 
+    for (ReadOnlyTStore.TStatus t : ReadOnlyTStore.TStatus.values()) {
+      MutableGaugeLong g = registry.newGauge(FATE_TX_STATE_METRIC_PREFIX + t.name().toUpperCase(),
+          "Transaction count for " + t.name() + " transactions", 0L);
+      fateTypeCounts.put(t.name(), g);
+    }
+  }
+
+  /**
+   * For testing only: allow refresh delay to be set to any value, over riding the enforced minimum.
+   *
+   * @param minimumRefreshDelay
+   *          set new min refresh value, in seconds.
+   */
+  void overrideRefresh(final long minimumRefreshDelay) {
+    long delay = Math.max(0, minimumRefreshDelay);
+    this.minimumRefreshDelay = TimeUnit.SECONDS.toMillis(delay);
   }
 
   @Override
   protected void prepareMetrics() {
-    FateMetricValues fateMetrics = metricValues.get();
+
     long now = System.currentTimeMillis();
+
     if ((lastUpdate + minimumRefreshDelay) < now) {
-      fateMetrics = FateMetricValues.updateFromZookeeper(context, fateMetrics);
-      metricValues.set(fateMetrics);
+      metricValues.set(updateFromZookeeper());
+
       lastUpdate = now;
-      // update individual gauges that are reported.
-      currentFateOps.set(fateMetrics.getCurrentFateOps());
-      zkChildFateOpsTotal.set(fateMetrics.getZkFateChildOpsTotal());
-      zkConnectionErrorsTotal.set(fateMetrics.getZkConnectionErrors());
+
+      recordValues();
     }
   }
```

Comment:
I'll look into reworking this - the intent was to provide thread safe access without blocking or synchronization (I'm not clear that what is there fully archives this).  The `prepareMetrics()` call is going to come from the metrics system(s) threads and it is unclear how various sinks will "share" results - or if there would be multiple calls, possibly concurrently.  Any call should produce an atomic view of the metrics when the call is made - that is, all values should be from the same update - it is not as important that concurrent calls receive the "same" values, just that they are internally consistent with the values gathered at a single point in time.   

It may be easier to reason about if I used synchronization and the current approach could be a premature optimization that is not necessary for the frequency that the metrics are collected.

Reply:
It may be useful to add a comment about why the locking is there.  Even if the comment just says the locking may be not be needed and was added as a precaution.

Adding locks in plugins that are called by an external system and the plugin calls back out to the external system does have a chance if causing deadlock.  I doubt that is the case here, but can't be sure w/o fully understanding the threading and locking model of the external system.

Reply:
Confirmed that moving it under the conditional does not change the metrics reported to the file sink - this will let the metrics system pick updates vs all (I believe that we use / want all metrics, changes or not, so not much difference.)